fsharp-backend/tests/Tests/BwdServer.Tests.fs

```diff
@@ -256,198 +249,216 @@ let t filename =
-                UTF8.ofBytesUnsafe actual.body
-              ),
-              LibExecution.DvalReprExternal.parseJson (
-                UTF8.ofBytesUnsafe expected.body
-              )
-            )
-          with
-          | e -> None
-
-        match asJson with
-        | Some (aJson, eJson) ->
-          let serialize (json : JsonDocument) =
-            LibExecution.DvalReprExternal.writePrettyJson json.WriteTo
-          Expect.equal
-            (actual.status, aHeaders, serialize aJson)
-            (expected.status, eHeaders, serialize eJson)
-            $"({server} as json)"
-        | None ->
-          match UTF8.ofBytesOpt actual.body, UTF8.ofBytesOpt expected.body with
-          | Some actualBody, Some expectedBody ->
-            Expect.equal
-              (actual.status, aHeaders, actualBody)
-              (expected.status, eHeaders, expectedBody)
-              $"({server} as string)"
-          | _ ->
-            Expect.equal
-              (actual.status, aHeaders, actual.body)
-              (expected.status, eHeaders, expected.body)
-              $"({server} as bytes)"
-      }
-
-    if skip then
-      skiptest $"underscore test - {name}"
+    return meta
+  }
+
+let normalizeActualHeaders (hs : (string * string) list) : (string * string) list =
+  hs
+  |> List.filterMap (fun (k, v) ->
+    match k, v with
+    | "Date", _ -> Some(k, "xxx, xx xxx xxxx xx:xx:xx xxx")
+    | "expires", _
+    | "Expires", _ -> Some(k, "xxx, xx xxx xxxx xx:xx:xx xxx")
+    | "x-darklang-execution-id", _ -> Some(k, "0123456789")
+    | "age", _
+    | "Age", _ -> None
+    | "X-GUploader-UploadID", _
+    | "x-guploader-uploadid", _ -> Some(k, "xxxx")
+    | "x-goog-generation", _ -> Some(k, "xxxx")
+    | other -> Some(k, v))
+  |> List.sortBy Tuple2.first // CLEANUP ocaml headers are sorted, inexplicably
+
+let normalizeExpectedHeaders
+  (headers : (string * string) list)
+  (actualBody : byte array)
+  : (string * string) list =
+  headers
+  |> List.map (fun (k, v) ->
+    match k, v with
+    // JSON can be different lengths, this plugs in the expected length
+    | "Content-Length", "LENGTH" -> (k, string actualBody.Length)
+    | _ -> (k, v))
+  |> List.sortBy Tuple2.first
+
+/// create a TCP client, used to make test HTTP requests
+let createClient (port : int) : Task<TcpClient> =
+  task {
+    let client = new TcpClient()
+
+    // Web server might not be loaded yet
+    let mutable connected = false
+    for i in 1..10 do
+      try
+        if not connected then
+          do! client.ConnectAsync("127.0.0.1", port)
+          connected <- true
+      with
+      | _ when i <> 10 ->
+        print $"Server not ready on port {port}, maybe retry"
+        do! System.Threading.Tasks.Task.Delay 1000
+    return client
+  }
+
+/// Makes the test request to one of the servers,
+/// testing the response matches expectations
+let runTestRequest canvasName testRequest testResponse server : Task<unit> =
```

Comment:
I find this helpful - knowing that (once setup is done), this is all that's required to run a test against one of the server.